### PR TITLE
gh-89640: Hardcode WASM float word ordering as little endian

### LINKED ISSUE #89640: configure script cannot detect float word ordering on linux
BPO | [45477](https://bugs.python.org/issue45477)
--- | :---
Nosy | @tiran
Files | <li>[config.log](https://bugs.python.org/file50364/config.log "Uploaded as text/plain at 2021-10-16.16:27:29 by sourish.basu")</li><li>[conftest.o](https://bugs.python.org/file50374/conftest.o "Uploaded as application/octet-stream at 2021-10-20.13:23:38 by sourish.basu")</li><li>[conftest.c](https://bugs.python.org/file50375/conftest.c "Uploaded as text/plain at 2021-10-20.13:23:53 by sourish.basu")</li><li>[conftest_unoptimized.o](https://bugs.python.org/file50376/conftest_unoptimized.o "Uploaded as application/octet-stream at 2021-10-20.14:08:43 by sourish.basu")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-10-14.21:19:53.235>
labels = ['build', 'expert-installation', '3.10']
title = 'configure script cannot detect float word ordering on linux'
updated_at = <Date 2021-10-22.22:49:31.555>
user = 'https://bugs.python.org/sourishbasu'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-22.22:49:31.555>
actor = 'sourish.basu'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Installation']
creation = <Date 2021-10-14.21:19:53.235>
creator = 'sourish.basu'
dependencies = []
files = ['50364', '50374', '50375', '50376']
hgrepos = []
issue_num = 45477
keywords = []
message_count = 12.0
messages = ['403951', '403978', '404096', '404135', '404138', '404447', '404448', '404449', '404450', '404459', '404463', '404836']
nosy_count = 2.0
nosy_names = ['christian.heimes', 'sourish.basu']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'compile error'
url = 'https://bugs.python.org/issue45477'
versions = ['Python 3.10']
```

</p></details>

### Linked PRs
* gh-125571
* gh-126387
* gh-126429
* gh-126430
* gh-126569
* gh-126747

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-89640

### CODE REVIEW

**Comment by hoodmane** [comment]
Thanks for the help @erlend-aasland!

**Comment by rruuaanng** [comment]
That looks good, but will fixing it to little endian cause other problems?

**Comment by hoodmane** [comment]
Any compliant implementation of webassembly is little endian:
https://webassembly.github.io/spec/core/binary/values.html#floating-point

I think that if someone forked the spec and invented "big endian wasm" but still called the target `wasm32-???-???` that would be met with a lot of hostility by the webassembly folks. Only wasm32-emscripten should ever hit this branch though, so it might be a tiny bit better to match `[*emscripten*]` and then we'll find out if `AX_C_FLOAT_WORDS_BIGENDIAN` ever started failing to detect wasi correctly, however unlikely that might be.

If for some reason we don't want to hard code Emscripten, there is an option here:
https://github.com/autoconf-archive/autoconf-archive/pull/316

File: configure.ac [hunk 1]
```diff
@@ -5946,28 +5946,26 @@ AS_VAR_IF([ac_cv_gcc_asm_for_x64], [yes], [
...
-   in ARM mixed-endian order (byte order 45670123)])
-fi
+AX_C_FLOAT_WORDS_BIGENDIAN(
+  [AC_DEFINE([DOUBLE_IS_BIG_ENDIAN_IEEE754], [1],
+             [Define if C doubles are 64-bit IEEE 754 binary format,
+              stored with the most significant byte first])],
+  [AC_DEFINE([DOUBLE_IS_LITTLE_ENDIAN_IEEE754], [1],
+             [Define if C doubles are 64-bit IEEE 754 binary format,
+              stored with the least significant byte first])],
+  [AS_CASE([$host_cpu],
```
Comment (erlend-aasland) [inline configure.ac:5956]: `host_cpu` is `wasm32` on Emscripten, so we cannot match directly against `*emscripten*`. Alternatively, we can rewrite the switch to check `$host/$host_cpu` and alter the ARM check to match `*/arm*` and the Emscripten check to match `*emscripten/*`. cc. @hoodmane

**Review by brettcannon — APPROVED** [review]

File: configure.ac (hunk 1 above)
Comment (hoodmane) [reply configure.ac:5956]: I think it's fine as is, particularly if @brettcannon approves

**Comment by hoodmane** [comment]
One effect that this change had is that it will no longer raise the error:
```
Unknown float word ordering. You need to manually preset
ax_cv_c_float_words_bigendian=no (or yes) according to your system.
```
if the float word order does not match *arm* or *wasm* and `AX_C_FLOAT_WORDS_BIGENDIAN` returns undefined. I think we should consider doing this:

```patch
--- a/configure.ac
+++ b/configure.ac
@@ -5920,7 +5920,15 @@ AX_C_FLOAT_WORDS_BIGENDIAN(
                                 stored in ARM mixed-endian order (byte order 45670123)])],
            [wasm*], [AC_DEFINE([DOUBLE_IS_LITTLE_ENDIAN_IEEE754], [1],
                                [Define if C doubles are 64-bit IEEE 754 binary format,
-                                stored with the least significant byte first])])])
+                                stored with the least significant byte first])],
+           AC_MSG_ERROR[([
+
+Unknown float word ordering. You need to manually preset
+ax_cv_c_float_words_bigendian=no (or yes) according to your system.
+
+           ])]
+  )]
+)
 
 # The short float repr introduced in Python 3.1 requires the
 # correctly-rounded string <-> double conversion functions from
```

**Comment by erlend-aasland** [comment]
Good catch, @hoodmane! Yes, indeed we should. Would you like to propose a PR?

**Pull request merged**